Migrate to the new OpenAssetIO unified entity data model

foundrytom opened the pull request:
Draft till https://github.com/TheFoundryVisionmongers/OpenAssetIO-MediaCreation/pull/2, https://github.com/TheFoundryVisionmongers/OpenAssetIO-MediaCreation/pull/3 and https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/363 are merged.

Once in, this switches to using the new trait-centric model for resolution.

foundrytom committed:
[Plugin] Switch to the new unified entity data model

This now requires the use of the `feature/181-coreCppBuild` branch
of OpenAssetIO.

As we’re still early doors with this, we’re happy that the process of
getting all this running is still somewhat manual. Once we have PyPy
packages published for `openassetio` and `openassetio_mediacreation`
this should get a lot simpler.

Signed-off-by: Tom Cowland <tom@foundry.com>

feltech commented on .github/workflows/ci.yaml:
```diff
@@ -39,25 +38,55 @@ jobs:
+        id: cache-conan
+        uses: actions/cache@v2
+        with:
+          path: ~/conan
```
[fix] It's a bit confusing, but `conan install --install-folder ~/conan ...` only tells conan where to put generated config files (e.g. CMake scripts).  By default the actual packages are installed in `~/.conan` (note the `.`).  

In the Vagrant bootstrap script(s) in OpenAssetIO there is a line `export CONAN_USER_HOME="$HOME/conan"` - that's the bit that tells conan where to install packages to.  In OpenAssetIO the packages and generated scripts are put in the same location, simply for convenience.

So we have a few options. The simplest is to just change this line to 
```suggestion
          path: ~/.conan
```
It means the generated files are not cached, but then they're probably regenerated every time anyway.

Or we could set the `CONAN_USER_HOME` environment variable, like in OpenAssetIO, to be explicit.

foundrytom replied:
Ahhhhhh, awesome thanks, I did wonder if that var was a DRY device, or actually also doing Conan things. I like your suggestion, given hopefully all this will go soon.

feltech commented on README.md:
```diff
@@ -61,14 +61,19 @@ python -m venv .venv
-pip install dependencies/OpenAssetIO
+git clone -b feature/181-coreCppBuild git@github.com:TheFoundryVisionmongers/OpenAssetIO.git dependencies/OpenAssetIO
+cmake -S dependencies/OpenAssetIO -B build
+cmake --build build --target openassetio-python-venv
```
[fix] This branch needs updating
```suggestion
cmake --build build --target openassetio-python-py-install
```

foundrytom replied:
🦅 👁️

feltech commented on README.md:
```diff
@@ -61,14 +61,19 @@ python -m venv .venv
 To use the linker in OpenTimelineIO `openassetio` needs to be installed.
-At present, this requires installation from source:
+At present, this requires installation from source of the
+`feature/181-coreCppBuild` branch.
```
[fix] OpenAssetIO-MediaCreation is also a dependency now

foundrytom replied:
Added [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/e3e92937a87978e2b8e3a52db8401d846f17ab6c).

feltech commented on otio_openassetio/operations/openassetio_media_linker.py:
```diff
@@ -36,19 +40,33 @@ def link_media_reference(in_clip, media_linker_argument_map):
+    ClipTrait(context.locale).setName(in_clip.name)
+
+    # In this simple implementation, we only need the URL to the media,
+    # so we use the LocatableContentTrait directly to avoid the manager
```
[consider] Might be worth adding a note "as opposed to all the traits of a particular specification"?

foundrytom replied:
Extended [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/ae293a72c290d407199b44744ae4c4f7bf5fae5a), and further corrected [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/400f641a4103b8357f49f6ac5ff4a0e64f23d3fb).

feltech commented on .github/workflows/ci.yaml:
```diff
@@ -39,25 +38,55 @@ jobs:
+          . dist/bin/activate
           pytest tests
         env:
           OPENASSETIO_PLUGIN_PATH: dependencies/OpenAssetIO/resources/examples/manager/BasicAssetLibrary/plugin
```
[question] Not part of this PR, but the `bal_plugin_env` pytest fixture overwrites the env var, so is this here for a particular reason (similarly in the README instructions)?

foundrytom replied:
You are 100% right there good sir. Addressed [here](https://github.com/TheFoundryVisionmongers/otio-openassetio/pull/5/commits/96c5e22038d08d5c8ad123294f31dc109954d025).

feltech reviewed (changes requested):
Took me a while to wrangle things to get tests running, but looks good and works well.  Just a couple of CI/docs fixes.

foundrytom committed:
[Docs] Remove unnecessary plugin search path instructions

As the test fixture sets these, this is not required - and actually overwritten.

Signed-off-by: Tom Cowland <tom@foundry.com>

feltech commented on README.md:
````diff
@@ -60,15 +60,29 @@ python -m venv .venv
+. dist/bin activate
 ```
 
+The Media Creation library is more straight forward.
````
[consider]
```suggestion
The Media Creation library is more straightforward.
```

feltech reviewed (approved):
Just a minor typo, feel free to ignore/fix then squash and merge.
